build/accelerate_launch.py

```diff
@@ -0,0 +1,111 @@
+"""
+
+# Standard
+import json
+import os
+import base64
+import pickle
+import logging
+
+# Third Party
+from accelerate.commands.launch import launch_command_parser, launch_command
+import torch
+
+
+def txt_to_obj(txt):
+    base64_bytes = txt.encode("ascii")
+    message_bytes = base64.b64decode(base64_bytes)
+    try:
+        # If the bytes represent JSON string
+        return json.loads(message_bytes)
+    except UnicodeDecodeError:
+        # Otherwise the bytes are a pickled python dictionary
+        return pickle.loads(message_bytes)
+
+
+def main():
+    LOGLEVEL = os.environ.get("LOG_LEVEL", "WARNING").upper()
+    logging.basicConfig(level=LOGLEVEL)
+
+    json_configs = {}
+    json_path = os.getenv("SFT_TRAINER_CONFIG_JSON_PATH")
+    json_env_var = os.getenv("SFT_TRAINER_CONFIG_JSON_ENV_VAR")
+
+    if json_path:
+        with open(json_path, "r", encoding="utf-8") as f:
+            json_configs = json.load(f)
+
+    elif json_env_var:
+        json_configs = txt_to_obj(json_env_var)
+
+    parser = launch_command_parser()
+    # determine which flags are store true actions (don't require a value to be set)
+    actions_type_map = {
+        action.dest: type(action).__name__ for action in parser._actions
+    }
+
+    # parse accelerate_launch_args
+    accelerate_launch_args = []
+    accelerate_config = json_configs.get("accelerate_launch_args", {})
+    if accelerate_config:
+        logging.info("Using accelerate_launch_args configs: %s", accelerate_config)
+        for key, val in accelerate_config.items():
+            if actions_type_map.get(key) == "_AppendAction":
+                for param_val in val:
+                    accelerate_launch_args.extend([f"--{key}", str(param_val)])
+            else:
+                accelerate_launch_args.append(f"--{key}")
+                # Only need to add key for params that aren't flags ie. --quiet
+                if actions_type_map.get(key) == "_StoreAction":
+                    accelerate_launch_args.append(str(val))
+
+    if json_configs.get("multi_gpu"):
+        # add FSDP config
+        if not accelerate_config.get("config_file"):
+            fsdp_filepath = os.getenv(
+                "FSDP_DEFAULTS_FILE_PATH", "/app/accelerate_fsdp_defaults.yaml"
+            )
+            if os.path.exists(fsdp_filepath):
+                logging.info("Setting accelerate config file to: %s", fsdp_filepath)
+                accelerate_launch_args.append("--config_file")
+                accelerate_launch_args.append(fsdp_filepath)
+
+            # add num_processes to overwrite config file set one
+            if not accelerate_config.get("num_processes"):
+                num_gpus = torch.cuda.device_count()
+                if num_gpus > 1:
+                    logging.info("Setting accelerate num processes to %s", num_gpus)
+                    accelerate_launch_args.append("--num_processes")
+                    accelerate_launch_args.append(str(num_gpus))
```

Comment:
Got confused here, do we intend to execute this block only when there is no `config_file` given?

Reply:
Should we not execute it when there is a config file from the user, in that case as well should not we have to prepare `accelerate launch` params isn't it?

Reply:
@anhuong sorry im unclear about this, because it is a valid use case that one will profile a `config_file` that contains all of the defaults, but while running experiments, wants to change one key without making a completely new defaults file.

Reply:
agreed that this coditional block needs to be refactored a bit. This should be the pseudocode-

```
if multi_gpu:
     if no config_file given:
            use default FSDP file
     if num_processes not provided by user:
         num_processes = num_gpus # to ensure it overrides the default file
else:
   #now we know its single GPU
   num_processes=1
```

for the case where it is multi_gpu and num_processes is provided by user, it should have been set in code above, in accelerate_launch_args already

Reply:
Thanks @Ssukriti Above pseudocode looks good to me, thank you!

Reply:
This thread is resolved once changes are pushed.
cc: @Ssukriti

Reply:
Thanks @tharapalanivel thats should be enough from my end. Thank you.

Reply:
My logic has it that if a config_file is passed by the user, we do not override with the default config_file or num_processes **we** set. However, we always passed all of the user provided arguments so that any individual field can be ovewritten in the config_file that a user passes and/or the default one set by us. 

So in my logic the user provided arguments is always passed in as `accelerate_launch_args`. If a user passes their own config_file, num_processes must be included for it to be valid.

Please correct me if i'm missing something

Reply:
Note that we have simplified this logic to put the onus on users to pass num_processes explicitly. This determines if running for multi-gpu or single gpu:
```
if num_processes set by user:
   if num_processes > 1:  # multi-gpu
      if not config_file:
         set FSDP config
   else if num_processes == 1:   # single-gpu
      os.environ[CUDA_VISIBLE_DEVICES] = "0"
else:   # num_processes not set
   warn num processes not passed in
   # this means num_processes will be pulled from config_file user provided or will be determined by accelerate launch
```